### PR TITLE
#1580 Support several AuthenticationProviderKey values in the Route

### LINKED ISSUE #1580: Support several AuthenticationProviderKey values in the Route
## New Feature
Opportunity to support several [AuthenticationProviderKey](https://ocelot.readthedocs.io/en/latest/features/authentication.html?highlight=AuthenticationProviderKey) values in the Route.

## Motivation for New Feature
At the moment, [Route](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Configuration/DownstreamRoute.cs) can contain one [AuthenticationProviderKey](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Configuration/AuthenticationOptions.cs#L12) value only. So, if we want to handle Route for different authentication providers, we have to specify two Routes and make Route path unique. Sometime it is not so easy to specify unique paths for the same endpoint.

## Useful Links
- [Features | Authentication](https://ocelot.readthedocs.io/en/latest/features/authentication.html) &rarr; ([authentication.rst](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/authentication.rst)), Ocelot docs
- [Authorize with a specific scheme in ASP.NET Core | Use multiple authentication schemes](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/limitingidentitybyscheme?view=aspnetcore-8.0#use-multiple-authentication-schemes), Microsoft Learn
- [Basic Authentication in C# / .NET](https://scatteredcode.net/basic-authentication-in-c-sharp-net/), ScatteredCode by Alexandru Puiu
- [.NET 6.0 - Basic Authentication Tutorial with Example API](https://jasonwatmore.com/post/2021/12/20/net-6-basic-authentication-tutorial-with-example-api), Jason Watmore's Blog

## Related
- #446 
- #1002

### PR DESCRIPTION
**Pull request by MayorSheFF**
## Closes #1580 
- #1580 

## Proposed Changes
  - opportunity to have the several authentication provider keys in Routes what allows to authenticate requests via different providers

### CODE REVIEW

**Comment by raman-m** [comment]
@MayorSheFF Igor,
Thanks for PR creation!
And, welcome to code review!

**Comment by raman-m** [comment]
@gdperkins @valleskon @clintonbale @sivapriyanc @SimeonBanev 
As an active participants of issue #1580 discussions, could you look at this solution please?
Your code review feedbacks are welcomed!

**Review by raman-m — CHANGES_REQUESTED** [review]
@MayorSheFF  Igor,
I see some added unit tests, this is very good,
But we have to update documentation to explain developers how to use this new feature.

File: src/Ocelot/Configuration/AuthenticationOptions.cs
Comment (RaynaldM) [inline src/Ocelot/Configuration/AuthenticationOptions.cs:19]: it's better to use an abstraction (e.g. IList) whenever possible

File: src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs
Comment (RaynaldM) [inline src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs:10]: maybe, it could be a real array

Comment (RaynaldM) [inline src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs:10]:
```suggestion
        private string[] _authenticationProviderKeys = new();

```

**Review by RaynaldM — APPROVED** [review]

File: src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs
Comment (raman-m) [reply src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs:10]: Yeah, it seems we're going to use arrays in [documentation](https://ocelot.readthedocs.io/en/latest/features/authentication.html?highlight=AuthenticationProviderKey).

**Review by RaynaldM — DISMISSED** [review]

**Comment by MayorSheFF** [comment]
> @MayorSheFF Igor, I see some added unit tests, this is very good. But we have to **update documentation** to explain developers how to use this new feature.

Hi @raman-m,

I am ready to do it if I have permissions. Please, explain to me where and how.

I apologize for not getting with you before now.

File: src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs
Comment (raman-m) [reply src/Ocelot/Configuration/Builder/AuthenticationOptionsBuilder.cs]: ✔️ Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1666/commits/10343db6558a697586e880f486e34909fdf1cd10

**Comment by MayorSheFF** [comment]
Hi Everyone,

Please, help to solve merge conflicts who has write access.

Thank you a lot.

**Comment by MayorSheFF** [comment]
Hi Everyone,

Could you please approve it if everything is good?

Thank you.

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
@MayorSheFF commented [last week](#issuecomment-1729826480):
> I am ready to do it if I have permissions. Please, explain to me where and how.

You have enough permissions as developer who code in his forked repository (current feature branch).
All documentation locates in the [docs](/MayorSheFF/MayorSheFF-Ocelot/tree/feature/1580-several-auth-options/docs) folder, as **.rst** files.
As an advice, it seems we have to review and update the following pages: [Search for AuthenticationProviderKey](https://ocelot.readthedocs.io/en/latest/search.html?q=AuthenticationProviderKey&check_keywords=yes&area=default)
I understand it will be hard for you to update docs, but at least you need to make a try.
Anyway I will update docs during final code review soon...

**Comment by raman-m** [comment]
> @MayorSheFF commented [on Sep 27](#issuecomment-1737234883)

Don't worry! It is [approved](#pullrequestreview-1646879314) by @RaynaldM 🤣 
But my approval will come after completing code review and updating docs... to perform merging to **develop** finally.
Be patient! Soon...

**Comment by raman-m** [comment]
@MayorSheFF Hi Igor!
How are you, man?
We've [discussed](../discussions/1771#discussioncomment-7486024) by the team, and this PR is included in [November 2023](../labels/Nov%2723) release. :yum:
Get prepared for pair programming! 😉

**Comment by MayorSheFF** [comment]
Hi, @raman-m, Raman.

Thank you a lot for good news. I apologize for not writing a doc until now. As usual for every developer, the main project just is overwhelming.  😉 Sometimes I check out emails and in several minutes I forget. Time management is not on a good level. 😉

I will try to find time this Thursday (or this weekend) to add the doc.

**Comment by raman-m** [comment]
> @MayorSheFF commented [on Nov 15](#issuecomment-1811313941)

Welcome back!

**Comment by MayorSheFF** [comment]
Hi @raman-m, Raman,

I added my description to the Authentication topic in the documentation.  I am not sure I chose a right place for that. There is "More identity providers" paragraph. Maybe, my changes should be there. If so, I can move it.

Please, review it and inform me of any deviations from project guidelines or if any corrections are needed.

Sorry for delay.

**Comment by raman-m** [comment]
@MayorSheFF commented [on Nov 18](#issuecomment-1817475281):
> There is "More identity providers" paragraph. Maybe, my changes should be there. 

**No**. Have you integrated new type of identity provider? It seems **No**.
New developer's instructions should be at the beginning of the doc (before "**JWT Tokens**" subsection)

I see two user scenarios:
- Developer uses only one auth-provider. We can wrap this scenario with old content as subsection, for example "Single Provider".
- Developer uses multiple auth-providers. We have to add new subsection with clean instructions, with possible link to samples. Named as "Multiple Providers"

We need to postpone writing docs for now! We have to develop final version of the feature first.

**Comment by raman-m** [comment]
I tried to review feature requirements and current implementation. You've added new `AuthenticationProviderKeys` property but the old `AuthenticationProviderKey` one remains. And what to do? It looks like breaking changes or not completed feature.

I see the following implementation plan:
- Expect multiple code reviews! Sorry, we have significant changes. You can even push @RaynaldM and @ggnaegi to review once again 😉 
- There are some unit tests. 🆗  What about acceptance tests? We have to write them If there are some acceptance tests then they test user scenarios of single provider. But we have to test multiple providers scenarios too.
- Integration tests aka End-to-end? I'm not sure. We could replace them with a sample. We can develop a authentication [sample](../tree/develop/samples) and show how to integrate. So, we need to develop a practical sample (gateway with routes + downstream apps with support of special auth-providers). The linked #1580 issue has good practical code snippets to reuse from other contributors & stakeholders.
- Update [docs](../tree/develop/docs) finally

I'm worried about backwards compatibility... If we introduced breaking changes then we must describe and add them to Release Notes. So, expect multiple reviews with long discussions.

@ggnaegi  FYI Feature is incomplete. Now I'm going to review the code after this 👆 requirements review...
November time is marching! 😉 Intensifying efforts... :smiley:

**Comment by MayorSheFF** [comment]
Hi, @raman-m, Raman.

> You've added new `AuthenticationProviderKeys` property but the old `AuthenticationProviderKey` one remains. And what to do?

It is just my point of view. I would suggest to leave both implementations: AuthenticationProviderKey and AuthenticationProviderKey**s**. It exactly gives backward compatibilities. As for me, it doesn't look like a big issue to have two implementations together. 

**But again, it is just my point of view.**

**Comment by raman-m** [comment]
> @MayorSheFF commented [on Nov 20](#issuecomment-1819030887)

Yes, we have to think in terms of backward compatibility. With high probability we will save the old property.

There are new commits in **develop** branch! Please, sync your fork, and rebase the feature branch please.

**Comment by MayorSheFF** [comment]
Hi, @raman-m.

> There are new commits in **develop** branch! Please, sync your fork, and rebase the feature branch please.

I've done it.

**Comment by raman-m** [comment]
Igor,
Get prepared for more active communication and pair programming next week please!
Your feature is included in [Nov'23](../milestone/2) milestone, so we will work more next week to complete the feature.

**Review by raman-m — CHANGES_REQUESTED** [review]
@MayorSheFF, we need to work on acceptance tests.
There are unit tests, but related acceptance test should be updated too.
Also, as a final check we need to add your user scenario acceptance test to ensure the feature works as you've requested/described.

Do you know how / have an intention to update acceptance tests?

**Comment by MayorSheFF** [comment]
> Do you know how / have an intention to update acceptance tests?

Hi, @raman-m.

Honestly, I have not thought about it yet. But... I will look at it. Should I do it this weekend or it can be done starting from Monday? I mean how urgent it is.

**Review by ggnaegi — CHANGES_REQUESTED** [review]
Hello
Thanks for the great PR, it's very useful. I've added some comments and it should clarify the behaviour when AuthenticationProviderKey and AuthenticationProviderKeys are set. Some unit tests and acceptance tests are missing, and the scenario with AuthenticationProviderKeys is not verified in my opinion.

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (ggnaegi) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:61]: Even better, we should throw if such configuration is defined...

**Comment by raman-m** [comment]
@MayorSheFF commented [on Dec 15](https://github.com/ThreeMammals/Ocelot/pull/1666#issuecomment-1858200020):
> Should I do it this weekend or it can be done starting from Monday? I mean **how urgent it is.**

Yeap, it is quite urgent :smile: because the PR is added to the current release. 
And we've agreed on pair programming [3 weeks ago](https://github.com/ThreeMammals/Ocelot/pull/1666#issuecomment-1826282626)...

I guess it is better to work on [code review issues](https://github.com/ThreeMammals/Ocelot/pull/1666#pullrequestreview-1784994000) now.
I can help you with current acceptance tests... Need to understand what we have now in relation to this feature

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (raman-m) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:61]: Throwing an exception means it will be caught by our lovely Error Mapper logic, right?

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (raman-m) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:67]: `route?.AuthenticationOptions?.AuthenticationProviderKeys is null` :question:

Comment (raman-m) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:67]: Idea for backward compatibility
```c#
if (route.AuthenticationOptions.AuthenticationProviderKeys is null)
{
    // Not Keys but Key for backward compatibility
    if (route.AuthenticationOptions.AuthenticationProviderKey)
    { // .... }
}
```

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (raman-m) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:79]: As I understood Igor's idea and design,
He is going to find the first key and validate, and authenticate and if Success then **break** the looping of all keys/providers.
Am I right, @MayorSheFF :grey_question:

File: src/Ocelot/Configuration/File/FileAuthenticationOptions.cs
Comment (raman-m) [reply src/Ocelot/Configuration/File/FileAuthenticationOptions.cs:11]: This is copy-constructor which was added in another PR #1753, it is not Igor's code :smile: 
So, copy-constructor is more performant than Reflection auto-mapping helpers.

Comment (raman-m) [reply src/Ocelot/Configuration/File/FileAuthenticationOptions.cs:11]: `from.AuthenticationProviderKeys`
Agree it should be remapped :wink:

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: @raman-m,

Yep, exactly.

@ggnaegi,

I've amended it.

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: @ggnaegi,

I've amended it.

Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: @raman-m,

Sorry. I didn't see your comments. So, I pushed my changes what @ggnaegi pointed out.

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: @ggnaegi,

Changed it too. But... I left `IEnumerable<>` to avoid unnecessary allocation by using `ToArray()`. Please, look at it.

**Comment by MayorSheFF** [comment]
I am going to take  look at Acceptance Tests. It wasn't included into the last commit.

**Comment by raman-m** [comment]
@ggnaegi @MayorSheFF  #1002
- #1002 

**FYI** 
Seems we have very closely related issue with a draft solution too...

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (ggnaegi) [inline src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:81]: What do you think of this example? I tend to try to reduce the nesting, but that's cosmetics...

```c#
var result = AuthenticateResult.NoResult();

if (route.AuthenticationOptions.AuthenticationProviderKeys.Length == 0)
{
    return result;
}

var authenticationProviderKeys =
    route
        .AuthenticationOptions
        .AuthenticationProviderKeys
        .Where(apk => !string.IsNullOrWhiteSpace(apk));

foreach (var authenticationProviderKey in authenticationProviderKeys)
{
    result = await httpContext.AuthenticateAsync(authenticationProviderKey);
    if (result.Succeeded)
    {
        return result;
    }
}

return result;
```

File: src/Ocelot/Configuration/File/FileAuthenticationOptions.cs
Comment (ggnaegi) [inline src/Ocelot/Configuration/File/FileAuthenticationOptions.cs:8]: not null :-)

**Review by ggnaegi — CHANGES_REQUESTED** [review]
@MayorSheFF, as @raman-m pointed out, the AuthenticationProviderKeys array will never be null. We can still improve the code a bit. And then, it's fine for me (even if you want to keep the nesting). Thanks!

**Comment by ggnaegi** [comment]
@MayorSheFF @raman-m I can help for the tests, just tell me what you want me to do...

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:61]: @ggnaegi, If you agree to allocation during every middleware execution, then for me it is not a problem to change. Just like this message. Thank you.

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: Resolved.

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: Resolved.

**Comment by MayorSheFF** [comment]
I've added changes what @ggnaegi had suggested.

I don't click "Resolve conversation". I don't know a rule. Should it be done by an owner of comments or by developer who solves it?

**Comment by ggnaegi** [comment]
> I've added changes what @ggnaegi had suggested.
> 
> I don't click "Resolve conversation". I don't know a rule. Should it be done by an owner of comments or by developer who solves it?

@MayorSheFF I agree with you about the null by the way, you are right. But that's not the point. You are the one implementing the feature, so, my understanding is, you click on resolve

File: src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs
Comment (ggnaegi) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs]: @MayorSheFF Yes, you are right, good point, so keep the null but declare result later?

Comment (MayorSheFF) [reply src/Ocelot/Authentication/Middleware/AuthenticationMiddleware.cs:71]: @ggnaegi, I brought it back. Please, check it out.

**Comment by ggnaegi** [comment]
@MayorSheFF Thanks!

**Comment by raman-m** [comment]
@ggnaegi 
Have the author resolved all your code review issues?

**Comment by ggnaegi** [comment]
> @ggnaegi 
> Have the author resolved all your code review issues?

Yes

**Comment by ggnaegi** [comment]
@raman-m why do you declare the result var in the loop? It might be helpful to return that the authentication has been rejected...

**Comment by raman-m** [comment]
@ggnaegi  Ups, Re-added the old definition & code.
But pay attention, it returns failed auth-result of the last key only which indicates all keys were failed.

**Comment by ggnaegi** [comment]
> @ggnaegi  Ups, Re-added the old definition & code.
> But pay attention, it returns failed auth-result of the last key only which indicates all keys were failed.
> 
> In our `Invoke` method we don't analyze the result much (no reasons) but we read `result.Principal` only for authenticated user.
> So, I don't see much difference between `AuthenticateResult.NoResult()` and failed result.
> 
> What will happen if `result.Principal` is null? We don't analyze it now.

Yes, but that's different than no result and for the unique authentication key, we will return succes or fail. Or we provide a custom fail result?

**Comment by raman-m** [comment]
> Or should we provide a custom error?

**Comment by raman-m** [comment]
Closed by mistake

**Comment by raman-m** [comment]
@MayorSheFF  Igor,
Given the current holiday weeks, it is not the best time to develop and contribute, but in any case I am not satisfied with the progress on this issue & PR...  I have been talking and communicating with you since June, more than 6 months!

Technically, it's stuck if we check the [development process](https://ocelot.readthedocs.io/en/latest/building/releaseprocess.html) on this PR... New features cannot be accepted and merged without acceptance tests + docs update. Writing unit tests is not sufficient for merging. The repo is public product being used by wide community of .NET engineers in the world. **The PR must have perfect quality!**

Still there are no any acceptance tests with your user scenarios from linked issue #1580  
As an author you should understand your user scenarios perfectly and you should be able to write tests easily... But there are no tests.

Considering the fact that the PR and issue were added to current upcoming release, I would say we delay the release, because the team agreed on adding it to the release.
I'm not happy with your occasional pair programming efforts to complete this feature. I will discuss this strange situation with development team...
As a team, we will try to finish this feature, but seems without your help. The proposed design will be confirmed by the team and we will use our vision and design of the feature.

**Comment by raman-m** [comment]
@ggnaegi  @RaynaldM 
By the way, this delays the release. We need to develop this feature ourselves...

**Comment by raman-m** [comment]
Require recreation...

@MayorSheFF  Igor,
Could you create new PR for this current feature branch **please ASAP**? ❗

**Comment by MayorSheFF** [comment]
Hi @raman-m,

I apologize for not getting in contact with you before now. I was in a business trip and now I am on vacation with my family. Unfortunately, I have a rule not to look at e-mails at all during vacation.

From last time, when Ocelot team had made a code review, I solved all comments. I didn't get any comments about teh documentation. Due to it, I left it as it is.

About the acceptance tests. Maybe I understood something wrong, but I thought that the acceptance tests would be covered by another member of the team. I don't remember exactly, but some people suggested to help with it and as I understood it was approved. Due to it, I didn't put anything there. Sorry for my misunderstanding, if it wasn't so.

Happy Christmas and happy New Year.

**Comment by MayorSheFF** [comment]
> @MayorSheFF, Could you create new PR for this current feature branch **please ASAP**? ❗ Thank you!

Is it still relevant?

**Comment by raman-m** [comment]
Please, press a few buttons to create new PR from this branch. Nothing more!..
The team will help you with acceptance tests.
And, enjoy your vacation!

**Comment by MayorSheFF** [comment]
Raman,

I have created the new PR #1870 .

**Comment by raman-m** [comment]
Thank you!

**Pull request closed without merging**
